### PR TITLE
Support adding fields to nested Parquet structs

### PR DESCRIPTION
**Pull request by billonahill**
Presto currently support adding columns to a table, but not adding fields to a nested struct. This adds support for struct field additions.

### CODE REVIEW

**Comment by zhenxiao** [comment]
@billonahill ,
@jxiang has a PR, also resolving adding/deleting/reordering fields in Parquet structs:
https://github.com/prestodb/presto/pull/4714

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java [hunk 1]
```diff
@@ -282,6 +262,61 @@ public ConnectorSplitSource getSplits(ConnectorTransactionHandle transaction, Co
...
+     * adding or dropping columns from the end of the table without modifying existing partitions is
+     * allowed, but every column that exists in both the table and partition must have the same type.
+     *
+     * If the type is a nested type (i.e. a struct), verify all partition struct fields exist in the
+     * table struct fields in with the same name, type and order. Af that condition is met, additional
+     * table struct fields are permitted.
+     *
+     * @throws PrestoException if the schema has not evolved in a supported way
+     */
+    private void verifySchemaEvolution(SchemaTableName tableName, List<FieldSchema> tableColumns,
```
Comment (dain) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java]: This check is used for all file formats, but it looks like only the Parquet record reader supports this.  What will happen with the other formats?

Comment (billonahill) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitManager.java]: Do we want to move forward with this PR or PR #4714 (or both)? If we want to merge this one, which could be an iteration towards a the larger #4714, then I can add integration tests to verify the other formats.

**Comment by jxiang** [comment]
I think it is better to have PR #4714 at first. If needed, then we can get this one.

**Comment by nezihyigitbasi** [comment]
@billonahill @zhenxiao just to confirm with you guys, does #4714 supersede this PR? If yes please close this one and then we can focus on the other.

**Comment by billonahill** [comment]
@nezihyigitbasi I've confirmed that this issue still exists on master by re-submitting this PR with only the test changes included. I will check against #4714 as well. Is there a way to exercise the `AbstractTestHiveClient.testTypesParquet` test without using travis CI or running docker locally? I'd like to be able to just run the unit test locally but `assertGetRecordsOptional` doesn't execute.

**Comment by electrum** [comment]
@billonahill The purpose of that test is to read data that was written by Hive. You can create the table in Hive, then run `TestHiveClient` with the parameters set to point at your Hive installation. The test table is created by running `create-test.sql` and `create-test-hive13.sql` in `presto-hive/src/test/sql` in Hive. Though for just that one test, you can modify the scripts to remove anything unnecessary (can take a long time if your Hive cluster is slow).

**Comment by electrum** [comment]
For IntelliJ, you set the parameters in the "Parameters" tab of the TestNG run configuration. I run those Hive tests all the time in the IDE.

**Comment by billonahill** [comment]
Thanks @electrum for jogging my memory re how to run these tests, it's been a while. :) Adding my own notes here, this is how I run from the command line:
```
mvn -pl presto-hive-hadoop2 -P test-hive-hadoop2 -Dhive.hadoop2.databaseName=default -Dtest=TestHiveClient -Dmaven.surefire.debug test
```

@nezihyigitbasi #4714 does not seem to address this issue. What's needed is for `HiveCoercionPolicy` to handle structs, as well as a `StructCoercionPolicy` in `HiveCoercionRecordCursor`. We could merge #4714 and I can follow up with a patch, or we can include the test changes from this patch in #4714 and implement a fix. Let me know what you and @jxiang prefer.

For my own notes, I've made a branch based on the patch in #4714 at https://github.com/billonahill/presto/tree/jxiang_extra_parquet_struct_fields which includes the fix to `HiveCoercionPolicy`, but still needs a fix to `HiveCoercionRecordCursor`.

**Comment by billonahill** [comment]
I've re-based the patch from master and updated it. This patch is ready to merge and can go in before #4714 (or #6675, which is much more extensive and failing with conflicts).

**Comment by zhenxiao** [comment]
@billonahill not so clear about this PR, could you please add more explanation. Is there anything missing in https://github.com/prestodb/presto/pull/4714 for schema evolution in Parquet? We just use that and are working fine

**Comment by billonahill** [comment]
@zhenxiao if you include the unit tests from this patch only in #4714 you can see that they fail. I didn't investigate the code in detail to see why though. The patch in #4714 is based on a pretty old version of the master and has conflicts so I might not have resolved correctly based on your intent. If you could rebase it and try running with the tests from this patch that would show if it works or not.

**Comment by nezihyigitbasi** [comment]
@zhenxiao can you give @billonahill 's suggestion a try and let us know if that test works or not?

**Comment by zhenxiao** [comment]
@nezihyigitbasi @billonahill I rebased our Nested Schema Evolution work at:
https://github.com/prestodb/presto/pull/6675

It becomes an independent PR for Nested Schema Evolution. I will close all other pending PRs for that.

@billonahill could you please take a test on 6675 to see whether that could fix the problem? We've using that in production for a long time, to resolve all schema evolution issues. It is quite stable.

**Comment by billonahill** [comment]
I've commented on #6675 that the problem this patch addresses still exists, even after applying the branch in #6675.

**Pull request closed without merging**